packages/playwright-core/src/server/injected/ariaSnapshot.ts

```diff
@@ -0,0 +1,240 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { InjectedScript } from './injectedScript';
+
+type AriaNode = {
+  role: string;
+  name?: string;
+  children?: (AriaNode | string)[];
+};
+
+export type AriaTemplateString = {
+  kind: 'string';
+  chunks: (RegExp | string)[];
+};
+
+export type AriaTemplateNode = {
+  kind: 'node';
+  role: string;
+  name?: string;
+  children?: (AriaTemplateNode | AriaTemplateString)[];
+};
+
+export function generateAriaTree(injectedScript: InjectedScript, rootElement?: Element): AriaNode {
+  const toAriaNode = (element: Element): { ariaNode: AriaNode, isLeaf: boolean } | null => {
+    const role = injectedScript.utils.getAriaRole(element);
+    if (!role)
+      return null;
+
+    const name = role ? injectedScript.utils.getElementAccessibleName(element, false) || undefined : undefined;
+    const isLeaf = leafRoles.has(role);
+    const result: AriaNode = { role };
+
+    if (isLeaf)
+      result.children = [name || element.textContent || ''];
```

Comment:
I am a bit worried about mixing name and text content in the same place. Wouldn't we end up with some nodes that have empty name, on purpose, and then dumping there text content for no reason? I am thinking about `aria-describedby` nodes or something similar.

Reply:
Agreed. I'll be able to set the name attr to an empty string to handle it later on. Leaving as is for now.